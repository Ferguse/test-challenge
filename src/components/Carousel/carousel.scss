$slick-font-path: "~slick-carousel/slick/fonts/";
$slick-loader-path: "~slick-carousel/slick/";

@import '~slick-carousel/slick/slick', '~slick-carousel/slick/slick-theme';
@import "../../styles/variables.scss";

.carousel {
  padding: 24px 0 42px;
  border-bottom: 1px solid $lightblue3;
  min-height: 400px;

  .slick-arrow::before {
    color: $lightblue3;
  }

  .slick-dots li, .slick-dots li button {
    width: 10px;
  }

  &_indicator {
    width: 20px;
    height: 20px;
    border-radius: 50%;
    border: 1px solid $lightblue3;
    display: inline-block;
    margin: 0 10px;
  }
}
