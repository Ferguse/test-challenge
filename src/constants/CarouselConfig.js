export const carouselConfig = {
  dots: true,
  infinite: true,
  speed: 500,
  slidesToShow: 5,
  slidesToScroll: 5,
  responsive: [
    {
      breakpoint: 1024,
      settings: {
        slidesToShow: 3,
        slidesToScroll: 3,
        arrows: false,
      },
    },
    {
      breakpoint: 600,
      settings: {
        slidesToShow: 2.5,
        slidesToScroll: 2,
        initialSlide: 1,
        arrows: false,
      },
    },
    {
      breakpoint: 480,
      settings: {
        slidesToShow: 1.3,
        slidesToScroll: 1,
        arrows: false,
      },
    },
  ],
};
