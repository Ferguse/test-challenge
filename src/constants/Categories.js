export const CATEGORIES = [
  { name: 'bike', icon: 'bike', color: '#ffa2a2', font: 'SourceSans' },
  { name: 'hiking', icon: 'hiking', color: '#a7fda7', font: 'Swansea' },
  { name: 'skating', icon: 'skating', color: '#a9a9fb', font: 'Goldleaf' },
];
