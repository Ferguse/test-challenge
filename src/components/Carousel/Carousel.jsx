import { useCallback, useMemo } from 'react';
import Slider from 'react-slick';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import { carouselConfig } from 'constants/CarouselConfig';
import { Card } from 'components/Card';
import { getCategory } from 'selectors/selectors';
import { setCurrentCategory } from 'actions';

import './carousel.scss';

const Carousel = ({ category }) => {
  const dispatch = useDispatch();
  const { name, color } = category;
  const books = useSelector(getCategory(name)) ?? {};

  const handleClick = useCallback(() => {
    dispatch(setCurrentCategory(name));
  }, [name]);

  const cards = useMemo(() => Object.values(books)
    .map((cardInfo) => <Card key={cardInfo.id} {...{ ...cardInfo, category }} />),
  [books]
  );

  return (
    <div className="carousel">
      <h2 className="title">
        <Link to={`/category/${name}`} onClick={handleClick}>{name}</Link>
        <i className="carousel_indicator" style={{ backgroundColor: color }}></i>
      </h2>
      <Slider {...carouselConfig}>
        {cards}
      </Slider>
    </div>
  );
};

Carousel.propTypes = {
  category: PropTypes.shape({
    name: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired,
  }).isRequired,
};

export { Carousel };
