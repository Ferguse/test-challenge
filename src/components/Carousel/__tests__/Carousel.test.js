import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockedStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import { CATEGORIES } from 'constants/Categories';

import { Carousel } from '../Carousel';

describe('components [Carousel]', () => {
  const state = {
    wishList: {},
    categories: {
      entities: [],
    },
    current: {},
  };

  const props = {
    category: CATEGORIES[0],
    volumeInfo: {
      publishedDate: '2020',
      title: 'test',
    },
    id: 'id',
  };

  const store = configureMockedStore([thunk])(state);
  const history = createMemoryHistory({ initialEntries: ['/'] });

  it('Renders', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router history={history} location={history.location} navigator={history}>
            <Carousel {...props} />
          </Router>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
