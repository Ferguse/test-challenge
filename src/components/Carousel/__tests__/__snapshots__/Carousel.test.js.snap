// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components [Carousel] Renders 1`] = `
<div
  className="carousel"
>
  <h2
    className="title"
  >
    <a
      href="/category/bike"
      onClick={[Function]}
    >
      bike
    </a>
    <i
      className="carousel_indicator"
      style={
        {
          "backgroundColor": "#ffa2a2",
        }
      }
    />
  </h2>
  <div
    className="slick-slider slick-initialized"
  >
    
    <div
      className="slick-list"
    >
      <div
        className="slick-track"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          {
            "WebkitTransform": "translate3d(0px, 0px, 0px)",
            "WebkitTransition": "",
            "msTransform": "translateX(0px)",
            "opacity": 1,
            "transform": "translate3d(0px, 0px, 0px)",
            "transition": "",
          }
        }
      />
    </div>
    
    
  </div>
</div>
`;
